keyring/backends/Windows.py

```diff
@@ -88,79 +301,152 @@ def priority(cls):
+        if not creds:
             return None
-        return res.value
+        password = creds.value
+        return password
 
-    def _get_password(self, target):
+    def _get_password_inner(self, target):
         try:
-            res = win32cred.CredRead(
-                Type=win32cred.CRED_TYPE_GENERIC, TargetName=target
-            )
-        except pywintypes.error as e:
-            e = OldPywinError.wrap(e)
-            if e.winerror == 1168 and e.funcname == 'CredRead':  # not found
+            res = CredRead(Type=CRED_TYPE_GENERIC, TargetName=target)
+        except CredError as e:
+            if e.winerror == 1168 and e.funcname == "CredRead":  # not found
                 return None
             raise
         return DecodingCredential(res)
 
+    def _get_password(self, target):
+        creds = self._get_password_inner(target)
+        if creds and ATTRIBUTE_KEYWORD in creds and creds[ATTRIBUTE_KEYWORD]["max_shards"] > 1:
+            for i in range(1, creds[ATTRIBUTE_KEYWORD]["max_shards"]):
+                shard = self._get_password_inner(TARGET_SHARD.format(target=target, n=i))
+                if not shard:
+                    log.critical(f"expected shard {i} not found; {creds[ATTRIBUTE_KEYWORD]}")
+                    return creds
+                else:
+                    creds["CredentialBlob"] += shard["CredentialBlob"]
+        return creds
+
     def set_password(self, service, username, password):
         existing_pw = self._get_password(service)
         if existing_pw:
-            # resave the existing password using a compound target
-            existing_username = existing_pw['UserName']
-            target = self._compound_name(existing_username, service)
-            self._set_password(
-                target,
-                existing_username,
-                existing_pw.value,
-            )
+            existing_username = existing_pw["UserName"]
+            if existing_username != username:
+                # resave the existing password using a compound target
+                target = self._compound_name(existing_username, service)
+                self._set_password(target, existing_username, existing_pw.value)
         self._set_password(service, username, str(password))
 
     def _set_password(self, target, username, password):
-        credential = dict(
-            Type=win32cred.CRED_TYPE_GENERIC,
-            TargetName=target,
-            UserName=username,
-            CredentialBlob=password,
-            Comment="Stored using python-keyring",
-            Persist=self.persist,
-        )
-        win32cred.CredWrite(credential, 0)
+        pwd_utf8_bytes = password.encode("utf-8")
+        pwd_len = len(pwd_utf8_bytes)
+
+        if pwd_len > MAX_PASSWORD_BYTES:
+            raise ValueError(MAX_PASSWORD_BYTES, "_set_password: {pwd_len=} exceeds {MAX_PASSWORD_BYTES=}")
+
+        n = MAX_PASSWORD_BYTES if self._is_init else self._max_password_bytes
```

Comment:
This logic doesn't smell right to me. Under what conditions is `bool(self._is_init) is False`?

Reply:
As commented in `__init__`,

```
        # prevent sharding during self._discover_max_password_bytes
        self._is_init = True
        self._max_password_bytes = self._discover_max_password_bytes()
        self._is_init = False
```

`_discover_max_password` returns the "measured" maximum length that fits in a Credentials Manager password record. `_discover_max_password` uses `_set_password` to "binary search" for the maximum size.  During that search, before we have the measured result, the search sets an arbitrary large `MAX_PASSWORD_BYTES` to bound the seach.  After we've measured the true "max", `self._is_init` is set to `False`, and the code simply uses the measured result (`self._max_password_bytes`).